### DISCUSSION TITLE
Spyder-project crashes after measurement procedure is finished

### DISCUSSION THREAD

**Discussion by HoWeBrz** [pymeasure/pymeasure#604] [General]
I am using Spyder5 in a Python39 virtual environment to create measurements using PyMeasure.
I basically modified the simple-procedure example with plot window (no managed window) to make a measurement from my oscilloscope (RigolDS1102E).
I am executing the python script from within Spyder from within an Windows Powershell.

After the measurement is done and the plot is shown properly (5 frequency measurements), mostly (somteimes not) I get an Python error-message: 
![image](https://user-images.githubusercontent.com/96414894/160087979-9d72353a-9df8-4c38-8d35-2132b621bd05.png)

After this error message occurs, the 
[Edit:] The IPython console gets restarted and  
Spyder does not find anyof my modules defined via PYTHONPATH.
I need to re-open the Spyder-project again to enable spyder to locate my modules.

[Solved:] (see below)
Additional issues are, that when I run the script multiple times (here in the screenshot I ran it 2 times) I always get the logging-output multiple times.
Here in the screenshot-logging window you see each log-output is printed 2 times - if I would have ran it 5 times, each log-ouput would be printed 5 times - Why ???
[:Solved]

Does anyone have an idea what happens ?

Attached is the python script.
[measFreq_100x_3s.py.txt](https://github.com/pymeasure/pymeasure/files/8348982/measFreq_100x_3s.py.txt)

**Comment by HoWeBrz** [comment]
Regarding this multiple log-ouput issue, here s a simple script to reproduce it:

```
from pymeasure.log import console_log
import logging
log = logging.getLogger(__name__)
log.addHandler(logging.NullHandler())
if __name__ == "__main__":
    console_log(log)
    log.info("Simple LogMessage")
```
![image](https://user-images.githubusercontent.com/96414894/160105684-5c4e3418-9696-4c5b-8476-6a01f5b3c2f8.png)

**Comment by LongnoseRob** [comment]
@HoWeBrz I know thir problem with the multiplied log messages..
I found that adding this to the end of my pymeasure scripts cleans up logging nicely.
(it might not be the best pyhtonic soluton, but it works.. )
```

handlers = log.handlers[:]
for handler in handlers:
    handler.close()
    log.removeHandler(handler)
```

**Reply by HoWeBrz** [reply]
Perfect ! 
That works to avoid multiple messages.

Any idea about the error message I get after the measurement is finished ?
Any idea how to debug this ?

**Reply by LongnoseRob** [reply]
> Any idea about the error message I get after the measurement is finished ?
> Any idea how to debug this ?

Does it also crashes if you run it from the command line?
There are some differences between a "normal" python and ipython session.
May be you could check if there are some settings in spyder for iphyton not to automagically restart and show a backtrace?

You also could try if the problem to isolate if the problem is  the procedure (running itwith out the plotter parts, just recording the data to a file) or from the plotter (running it without the procedure (reading data from a file).

**Reply by HoWeBrz** [reply]
I was able to isolate the problem to the "plotter" method from pyMeasure.
It appears like if something times out after 10-15 seconds and then crashes the ipython-kernel.
If I do not plot the results, the ipython-console never crashes.

I am wondering if I need to properly shutdown the plotter-gui, or so ?
How could this be done ?

**Comment by HoWeBrz** [comment]
When using "ipython" and "python" standalone (without Spyder), 
it happens that "ipython" crashes in the same way.

With "python" it does not crash, but gives this message after I close the plotter-gui:
```
11:55:03 AM: Finished the measurement (__main__, INFO)
QObject::~QObject: Timers cannot be stopped from another thread
```

There seems to be an issue with "ipython" and the timers used for the plotter ?

When googling for the above message, it seems that the termination of the plotter gui is not properly done, or that another thread tried to terminate the plotter-thread - whic is apparently not allowed in QT.
I would leave the investigations to the pyMeasure QT-experts.

**Reply by msmttchr** [reply]
I think this may be relate to https://github.com/pymeasure/pymeasure/issues/422.
We know that `Plotter` class is broken, but I am not sure this is your problem.

**Reply by HoWeBrz** [reply]
Would make sense that this is related, as it is stated in https://github.com/pymeasure/pymeasure/issues/422 that the way plotter is called is not according QT documentation.
For my application the plotter-class would be the better fit than the managed-windows. 
I hope it will be redesigned soon.
I will use the managed-window solution in the meantime.

**Comment by HoWeBrz** [comment]
Considered as answered, since there is an issue in the QT package.
